packages/api-client/src/main/tcp/WebSocketClient.test.node.js

```diff
@@ -35,75 +34,61 @@ const fakeHttpClient = {
   refreshAccessToken: () => Promise.resolve(accessTokenPayload),
 };
 
-const WEBSOCKET_PORT = 8087;
-const WEBSOCKET_URL = `ws://localhost:${WEBSOCKET_PORT}`;
-let server = undefined;
+describe('WebSocketClient', () => {
+  describe('handler', () => {
+    it('calls "onOpen" when WebSocket opens', async () => {
+      const websocketClient = new WebSocketClient('url', fakeHttpClient);
+      const onOpenSpy = spyOn(websocketClient, 'onOpen').and.callThrough();
+      const fakeSocket = {};
+      const socket = websocketClient.socket;
+      spyOn(socket, 'getReconnectingWebsocket').and.returnValue(fakeSocket);
+
+      await websocketClient.connect();
+      fakeSocket.onopen();
+
+      expect(onOpenSpy.calls.count()).toBe(1);
+    });
 
-function startEchoServer() {
-  server = new WebSocketServer({port: WEBSOCKET_PORT});
-  server.on('connection', ws => {
-    ws.on('message', message => {
-      server.clients.forEach(client => {
-        const payload = {
-          fromServer: `Echo: ${message}`,
-        };
+    it('calls "onClose" when WebSocket closes', async () => {
+      const websocketClient = new WebSocketClient('url', fakeHttpClient);
+      const onCloseSpy = spyOn(websocketClient, 'onClose').and.callThrough();
+      const fakeSocket = {};
+      const socket = websocketClient.socket;
+      spyOn(socket, 'getReconnectingWebsocket').and.returnValue(fakeSocket);
 
-        const options = {
-          binary: true,
-          mask: false,
-        };
+      await websocketClient.connect();
+      fakeSocket.onclose();
 
-        client.send(JSON.stringify(payload), options);
-      });
+      expect(onCloseSpy.calls.count()).toBe(1);
     });
-  });
 
-  server.on('error', error => console.error(`Echo WebSocket server error: "${error.message}"`));
-}
+    it('calls "onError" when WebSocket received error', async () => {
+      const websocketClient = new WebSocketClient('url', fakeHttpClient);
+      const onErrorSpy = spyOn(websocketClient, 'onError').and.callThrough();
+      const refreshTokenSpy = spyOn(websocketClient, 'refreshAccessToken').and.callThrough();
+      const fakeSocket = {};
+      const socket = websocketClient.socket;
+      spyOn(socket, 'getReconnectingWebsocket').and.returnValue(fakeSocket);
 
-describe('WebSocketClient', () => {
-  describe('"connect"', () => {
-    beforeEach(() => startEchoServer());
+      await websocketClient.connect();
+      fakeSocket.onerror(new Error('error'));
```

Comment:
The error is not thrown?

Reply:
No, because this is the call of the handler. In the actual code something would have catched and passed this error. (It's only an argument here.)